### PR TITLE
Refactor `Distributor.push()`

### PR DESCRIPTION
**Pull request by colega**
#### What this PR does

Makes `Distributor.push` shorter and hopefully cleaner.

#### Which issue(s) this PR fixes or relates to

None.

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -1378,25 +1341,52 @@ func (d *Distributor) push(ctx context.Context, pushReq *Request) error {
...
 }
 
-func preallocSliceIfNeeded[T any](size int) []T {
-	if size > 0 {
-		return make([]T, 0, size)
-	}
-	return nil
+// getSeriesAndMetadataTokensForBatchRequest returns a slice of tokens for the series and metadata from the request in this specific order.
+// Metadata tokens start at initialMetadataIndex.
+func getSeriesAndMetadataTokensForBatchRequest(userID string, req *mimirpb.WriteRequest) (keys []uint32, initialMetadataIndex int) {
```
Comment (bboreham) [inline pkg/distributor/distributor.go:1346]: What makes it a "BatchRequest" ?  This is called with the entire request; I thought the batches were what DoBatch splits it into.

**Review by bboreham — APPROVED** [review]
Seems ok; I have a question about naming.

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (colega) [reply pkg/distributor/distributor.go:1346]: My reasoning was: _since we pass this to ring.DoBatch_ ... but if this raised you a question, then it's not helping. I'm going to remove the `ForBatchRequest` suffix from this method.

File: pkg/mimirpb/timeseries.go [hunk 2]
```diff
@@ -590,3 +590,40 @@ func copyToYoloString(buf []byte, src string) (string, []byte) {
...
+// It assumes the indexes before the initialMetadataIndex are timeseries, and the rest are metadata.
+func (p *WriteRequest) ForIndexes(indexes []int, initialMetadataIndex int) *WriteRequest {
+	var timeseriesCount, metadataCount int
+	for _, i := range indexes {
+		if i >= initialMetadataIndex {
+			metadataCount++
+		} else {
+			timeseriesCount++
+		}
+	}
```
Comment (dimitarvdimitrov) [inline pkg/mimirpb/timeseries.go:604]: can't we do this with binary search?

**Review by dimitarvdimitrov — APPROVED** [review]
If you want to further simplify i think we can, but it looks good as-is too. Thanks for the cleanup

File: pkg/mimirpb/timeseries.go (hunk 2 above)
Comment (colega) [reply pkg/mimirpb/timeseries.go:604]: We could, but it's not necessarily faster. We'd need a benchmark, and I need to stop somewhere :D

**Pull request merged**